### DISCUSSION TITLE
Is there a way to enable minion on master?

### DISCUSSION THREAD

**Discussion by abhi1693** [cdalvaro/docker-salt-master#235] [Q&A] [answered]
I need to be able to fetch basic information about the master via the API such as version it's running. This is needed by 3rd party integrations to ensure we can begin the minion upgrade process when this version changes. One idea was to simply install a minion inside the master. I tried adding a `minion.conf` in the `config` directory with just `master: localhost` and it started throwing errors about missing directories.

I was wondering if this is a hidden feature of this repo so that I can simply enable this via environment variable and start managing our master too?

**Comment by abhi1693** [comment]
I managed to install `salt-minion` with the same version as master inside the container. But I had to update the `/etc/salt/minion` and `/etc/salt/minion_id` manually to make it work. It may be useful for others to get this functionality as well as part of the container runtime.

The container can create minimalistic configuration so that it works out-of-the-box by providing a way to overide the `minion_id`. I guess it can default to `local` when not set exlpicitly. The custom configuration can be modelled just like how master config is managed.

**Reply by cdalvaro** [reply]
Hi @abhi1693, I'm curious about this.

How can you use a salt-minion to get the salt-master version? And, what other things do you want to do with a minion inside the container?

It's just curiosity. I think I can add the this support if it adds value to the image. But I would like to know more details to do it right.

**Reply by abhi1693** [reply]
Via salt-api, I can send `test.version` targetting the minion running on the master. If this package is installed with the exact versoin of master, it's safe to assume the the version returned is of the master. Another way is to call the API with `salt-call --local test.version` using `cmd.run`.
Another thing, we want to do is grab the log files when we need to troubleshoot issues remotely rather than logging into the machine. Currently, these are the only 2 use cases I have but we may have more in future.

**Reply by cdalvaro** [reply]
There are some [modules for external logging](https://docs.saltproject.io/en/latest/ref/configuration/logging/handlers/index.html#external-logging-handlers).

Also you can [use rsyslog](https://docs.saltproject.io/en/latest/ref/configuration/logging/index.html#available-configuration-settings) to receive master's logs.

Having a `salt-minion` instance running inside the container will increase resources consumption and will add more complexity to the image (more things to fail)

I just want to have a good reason to add this support.

**Reply by cdalvaro** [reply, marked as answer]
Hi @abhi1693,

I have just released image `3007.0_2` with support for internal `salt-minion`.

Could you tell me the new release works fine for you?
